exonum-java-binding-common/src/main/java/com/exonum/binding/common/message/TransactionMessage.java

```diff
@@ -0,0 +1,200 @@
+  byte[] getPayload();
+
+  /**
+   * Returns the transaction message hash.
+   */
+  HashCode hash();
+
+  /**
+   * Returns the <a href="https://ed25519.cr.yp.to/">Ed25519</a> signature
+   * over this binary message.
+   *
+   * <p>The signature is <strong>not</strong> guaranteed to be valid and must be verified against
+   * the signer’s public key.
+   *
+   * @see CryptoFunctions#ed25519()
+   */
+  byte[] getSignature();
+
+  /**
+   * Returns the transaction message in binary format.
+   */
+  byte[] toBytes();
+
+  /**
+   * Returns the transaction message size in bytes.
+   */
+  int size();
+
+  /**
+   * Create a new builder for the transaction message.
+   */
+  static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Creates the transaction message from the given bytes array.
+   */
+  static TransactionMessage fromBytes(byte[] bytes) {
+    return new BinaryTransactionMessage(bytes);
+  }
+
+  /**
+   * Creates the transaction message from the given bytes buffer.
+   */
+  static TransactionMessage fromBuffer(ByteBuffer buffer) {
+    return new BinaryTransactionMessage(buffer.array());
+  }
+
+  /**
+   * Builder for the binary transaction message.
+   */
+  class Builder {
+    private Short serviceId;
+    private Short transactionId;
+    private ByteBuffer payload;
+
+    /**
+     * Sets service identifier to the transaction message.
+     */
+    public Builder serviceId(short serviceId) {
+      this.serviceId = serviceId;
+      return this;
+    }
+
+    /**
+     * Sets transaction identifier to the transaction message.
+     */
+    public Builder transactionId(short transactionId) {
+      this.transactionId = transactionId;
+      return this;
+    }
+
+    /**
+     * Sets payload to the transaction message.
+     */
+    public Builder payload(byte[] payload) {
+      return payload(ByteBuffer.wrap(payload));
+    }
+
+    /**
+     * Sets payload to the transaction message.
+     */
+    public Builder payload(ByteBuffer payload) {
+      this.payload = payload.duplicate().order(ByteOrder.LITTLE_ENDIAN);
+      return this;
+    }
+
+    /**
+     * Signs the message, creating a new signed binary transaction message.
+     *
+     * @param keys key pair with private and public keys. Public key is used as an author key of the
+     *        message and private key is used for signing the message.
+     * @param crypto a cryptographic function to use
+     * @return a new signed binary transaction message
+     * @throws NullPointerException if serviceId or transactionId or payload weren't set
+     * @throws IllegalArgumentException if public key has wrong size
+     */
+    public TransactionMessage sign(KeyPair keys, CryptoFunction crypto) {
+      checkRequiredFieldsSet();
+      PublicKey authorPublicKey = keys.getPublicKey();
+      checkArgument(authorPublicKey.size() == AUTHOR_PUBLIC_KEY_SIZE);
+
+      ByteBuffer buffer = ByteBuffer
+          .allocate(MIN_MESSAGE_SIZE + payload.limit())
```

Comment:
This code is incorrect: payload.limit() is not always the number of bytes that would be transferred in
ByteBuffer#put(ByteBuffer). 

Please add a test for that.

Reply:
I see, thanks. Fixed